### PR TITLE
Add STEP geometry parameterisations

### PR DESCRIPTION
**Pull request by shaildesai96**
## Linked Issues

Does this PR close or fix any Issues? Remember to create an Issue before starting work so that your fix / proposal can be addressed by the team.

e.g. Closes #{ID}

## Description

What is your PR trying to achieve? How did you go about achieving it?

## Interface Changes

If you've had to update an interface or introduce a new interface as part of your change then let us know here.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `flake8` and `black .`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by sebkahn** [comment]
Is this PR addresses https://github.com/Fusion-Power-Plant-Framework/bluemira/issues/179 ?

File: bluemira/geometry/parameterisations.py [hunk 1]
```diff
@@ -1044,3 +1044,345 @@ def create_shape(self, label=""):
...
+
+
+class SCCurvedPictureFrame(GeometryParameterisation):
+    """
+    Curved (Superconducting) picture-frame geometry parameterisation.
+    """
+
+    __slots__ = ()
+
+    def __init__(self, var_dict={}):
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/geometry/parameterisations.py:1774]: Please address this comment on all of your new parameterisations

File: tests/bluemira/geometry/test_parameterisations.py [hunk 2]
```diff
@@ -198,9 +199,17 @@ def test_segments(self):
...
         assert len(wire._boundary) == 4
-        p.adjust_variable("r", value=0)
+        p.adjust_variable("ri", value=0)
+        p.adjust_variable("ro", value=0)
         wire = p.create_shape()
-        assert len(wire._boundary) == 2
+        assert len(wire._boundary) == 4
+
+
+class TestCurvedPictureFrame:
```
Comment (CoronelBuendia) [inline tests/bluemira/geometry/test_parameterisations.py:208]: We should have one of these for each of these new parameterisations.

File: bluemira/geometry/parameterisations.py [hunk 3]
```diff
@@ -1146,28 +1293,28 @@ def __init__(self, var_dict=None):
...
+                    "ri",
+                    0.0,
+                    lower_bound=0,
+                    upper_bound=0.2,
+                    descr="Inboard corner radius",
+                ),
+                BoundedVariable(
+                    "ro", 2, lower_bound=0, upper_bound=5, descr="Outboard corner radius"
                 ),
             ],
```
Comment (CoronelBuendia) [inline bluemira/geometry/parameterisations.py:1319]: This changes the API to this parameterisation. Please rebase with develop and check that you don't break any of the tests that might use this parameterisation with its old interface.

**Review by CoronelBuendia — CHANGES_REQUESTED** [review]

File: tests/bluemira/geometry/test_parameterisations.py [hunk 4]
```diff
@@ -27,14 +27,21 @@
 import numpy as np
 import pytest
 
+from bluemira.display import plot_2d
```
Comment (CoronelBuendia) [inline tests/bluemira/geometry/test_parameterisations.py:30]: We try not to plot in tests any more. I appreciate they're useful, but it makes the tests a pain to run. Remove this import and the plot calls

**Review by CoronelBuendia — CHANGES_REQUESTED** [review]
Just a final one and looks good

**Review by CoronelBuendia — APPROVED** [review]

File: bluemira/geometry/parameterisations.py [hunk 5]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+class PictureFrameTools:
+    """
+    Curved Pictureframe parameterisation with straight (flat) inner limb
+
+    Parameters
+    ----------
+    var_dict: Optional[dict]
+        Dictionary with which to update the default values of the parameterisation.
+
+    """
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:998]: `PictureFrameTools` is not a inheriting from `GeometryParametrization` and does not have an `__init__` finction that would contain `var_dict`, are you sure this description is meant for `PictureFrameTools`?

File: bluemira/geometry/parameterisations.py [hunk 6]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        flip=False,
+        r_c=0,
+    ):
+        """
+        Makes smooth dome for CP coils
+
+        Returns
+        -------
+        shape: BluemiraWire
+            CAD Wire of the geometry
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1017]: Can we be more specific?

File: bluemira/geometry/parameterisations.py [hunk 7]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+
+        """
+        # Define basic Top Curve (with no joint or corner transitions)
+
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1021]:
```suggestion
```

File: bluemira/geometry/parameterisations.py [hunk 8]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        # Define the basic main curve (with no joint or transitions curves)
+        alpha = np.arctan(0.5 * (x_out - x_curve_start) / abs(z_top - z_mid))
+        theta_leg_basic = 2 * (np.pi - 2 * alpha)
+        r_leg = 0.5 * (x_out - x_curve_start) / np.sin(theta_leg_basic / 2)
+        z_top_r_leg = z_top + r_leg if flip else z_top - r_leg
+        leg_centre = (x_out - 0.5 * (x_out - x_curve_start), 0, z_top_r_leg)
+
+        # Transitioning Curve
+        sin_a = np.sin(theta_leg_basic / 2)
+        cos_a = np.cos(theta_leg_basic / 2)
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1033]: Detail: multiplication is always faster than divisions (for all the `/ 2` before an after this line)

```suggestion
        cos_a = np.cos(0.5 * theta_leg_basic)
```

File: bluemira/geometry/parameterisations.py [hunk 9]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        x_out,
+        x_curve_start,
+        x_mid,
+        z_top,
+        z_mid,
+        flip=False,
+        r_c=0,
+    ):
+        """
+        Makes smooth dome for CP coils
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1012]: Add the trigo equation in description?

Also interresting to explain the geometry build logic here (we start by this section, we add this act and then this other arc etc...)

File: bluemira/geometry/parameterisations.py [hunk 10]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1013]: Add the inputs description

``` python
"""
ARGUMENTS
---------
axis: ??
     This is my input axis
x_out: float
      This is my x_out
...
"""
```

File: bluemira/geometry/parameterisations.py [hunk 11]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        leg_centre = (x_out - 0.5 * (x_out - x_curve_start), 0, z_top_r_leg)
+
+        # Transitioning Curve
+        sin_a = np.sin(theta_leg_basic / 2)
+        cos_a = np.cos(theta_leg_basic / 2)
+        alpha_leg = (
+            np.arcsin(np.abs(r_leg * sin_a - r_c) / (r_leg - r_c)) + theta_leg_basic / 2
+        )
+
+        # Joint Curve
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1038]: Confusing comment:      
Are you taking about a true magnet joing or a section of the shape joining two elements?

File: bluemira/geometry/parameterisations.py [hunk 12]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+
+        # Build straight section of leg
+        p1 = [x_mid, 0, z_mid]
+        p2 = [leg_centre[0] - (r_leg + r_j) * np.sin(theta_j), 0, z_mid]
+
+        if not flip:
+            out_wire.append(make_polygon([p1, p2]))
+        else:
+            out_wire.append(make_polygon([p2, p1]))
+        out_wire.append(
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1059]: Is is still the straight section?

Seems to be the curve joining the main dome with the straight section (please add a `# Arc joining the straight section with the dome arc`)

File: bluemira/geometry/parameterisations.py [hunk 13]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+            make_circle(
+                radius=r_j,
+                center=joint_curve_centre,
+                start_angle=90 - np.rad2deg(theta_j) if flip else -90,
+                end_angle=90 if flip else np.rad2deg(theta_j) - 90,
+                axis=axis,
+                label="bottom_limb_joint" if flip else "top_limb_joint",
+            )
+        )
+
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1069]: Add description 

`# Bottom limb dome`

File: bluemira/geometry/parameterisations.py [hunk 14]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+            0,
+            z_mid_r_j,
+        )
+
+        # Final total angular extent of main curve
+        theta_leg_final = alpha_leg - (theta_leg_basic / 2 - theta_j)
+
+        # Build straight section of leg
+        p1 = [x_mid, 0, z_mid]
+        p2 = [leg_centre[0] - (r_leg + r_j) * np.sin(theta_j), 0, z_mid]
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1053]: I am never keen on variables diferenciated by numbers `p1` `p2`, but this is relatively ok here

File: bluemira/geometry/parameterisations.py [hunk 15]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        )
+
+        label = "bot_limb_dome" if flip else "top_limb_dome"
+
+        return BluemiraWire(out_wire, label=label)
+
+    @staticmethod
+    def _make_flat_leg(axis, x_in, x_out, z, r_i, r_o, flip=False):
+        """
+        Makes a flat leg with the option of one end rounded
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1090]: Which part of the leg is flat, the top (horizontal) or the outboard (vertical) ?

Also specify the arguments

File: bluemira/geometry/parameterisations.py [hunk 16]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+            wires.append(
+                make_circle(
+                    r_i,
+                    c_i,
+                    start_angle=180 if flip else 90,
+                    end_angle=270 if flip else 180,
+                    axis=axis,
+                    label="inner_bot_corner" if flip else "inner_top_corner",
+                )
+            )
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1113]: Add a empty line

File: bluemira/geometry/parameterisations.py [hunk 17]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        # Straight Section
+        p1 = [x_in + r_i, 0.0, z]
+        p2 = [x_out - r_o, 0.0, z]
+        wires.append(make_polygon([p1, p2], label="bot_limb" if flip else "top_limb"))
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1117]: Add an empty line

File: bluemira/geometry/parameterisations.py [hunk 18]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+            )
+        label = "bot_limb" if flip else "top_limb"
+
+        return BluemiraWire(wires, label=label)
+
+    @staticmethod
+    def _make_tapered_inner_leg(axis, x_in, x_mid, z_in, z_mid_up, z_mid_down):
+        """
+        Makes a tapered inboard leg using a circle arc taper
+
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1138]: Add arguments

File: bluemira/geometry/parameterisations.py [hunk 19]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+        Returns
+        -------
+        shape: BluemiraWire
+            CAD Wire of the geometry
+        """
+        p1 = [x_mid, 0, z_in]
+        p2 = [x_mid, 0, z_mid_up]
+        wires = [make_polygon([p1, p2], label="inner_limb_mid_up")]
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1146]: Add empty line

File: bluemira/geometry/parameterisations.py [hunk 20]
```diff
@@ -1046,74 +1231,36 @@ def create_shape(self, label=""):
...
-            )
+        top_leg = PictureFrameTools._make_flat_leg(
+            axis,
+            x1,
+            x2,
+            z1,
+            ri,
+            ro,
+            flip=False,
+        )
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1249]: Make it one line :

``` python 
        top_leg = PictureFrameTools._make_flat_leg(axis, x1, x2, z1, ri, ro, flip=False)
```

File: bluemira/geometry/parameterisations.py [hunk 21]
```diff
@@ -1046,74 +1231,36 @@ def create_shape(self, label=""):
...
-
+        bot_leg = PictureFrameTools._make_flat_leg(
+            axis,
+            x1,
+            x2,
+            z2,
+            ri,
+            ro,
+            flip=True,
+        )
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1262]: Make it one line

``` python
        bot_leg = PictureFrameTools._make_flat_leg(axis, x1, x2, z2, ri, ro, flip=True)
```

File: bluemira/geometry/parameterisations.py [hunk 22]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
...
+
+    Parameters
+    ----------
+    var_dict: Optional[dict]
+        Dictionary with which to update the default values of the parameterisation.
+
+    """
+
+    @staticmethod
+    def _make_domed_leg(
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1001]: Also good to specify types in declaration (for all function declaration on this PR)

``` python
       axis,
        x_out: float,
        x_curve_start: float,
        x_mid: float,
        z_top: float,
        z_mid: float,
        flip=False,
        r_c=0,
    ):
```

File: bluemira/geometry/parameterisations.py [hunk 23]
```diff
@@ -1189,43 +1336,633 @@ def create_shape(self, label=""):
...
+
         axis = [0, -1, 0]
+        wires = []
+        inb_leg = PictureFrameTools._make_tapered_inner_leg(
+            axis, x1, x2, z1, z2 - ri, -z2 + ri
+        )
+        wires.append(inb_leg)
 
-        wires = [make_polygon([p1, p2, p3, p4, p5, p6, p7, p8], label="inner_limb")]
+        top_leg = PictureFrameTools._make_flat_leg(
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1350]: One line

File: bluemira/geometry/parameterisations.py [hunk 24]
```diff
@@ -1189,43 +1336,633 @@ def create_shape(self, label=""):
...
-                    start_angle=270,
-                    end_angle=360,
-                    axis=axis,
-                    label="lower_corner",
-                )
-            )
+        wires.append(make_polygon([p5, p6], label="outer_limb"))
 
-        wires.append(make_polygon([p9, p10], label="outer_limb"))
+        bot_leg = PictureFrameTools._make_flat_leg(
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1363]: One line

File: bluemira/geometry/parameterisations.py [hunk 25]
```diff
@@ -1189,43 +1336,633 @@ def create_shape(self, label=""):
...
+        return BluemiraWire(wires, label=label)
 
-        if r != 0.0:
-            wires.append(
-                make_circle(
-                    r, c2, start_angle=0, end_angle=90, axis=axis, label="upper_corner"
-                )
-            )
+
+class FullDomeFlatInnerCurvedPictureFrame(GeometryParameterisation):
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1376]: Optional: This name is soooooo long, it it possible to reduce the name length?

File: bluemira/geometry/parameterisations.py [hunk 26]
```diff
@@ -1189,43 +1336,633 @@ def create_shape(self, label=""):
...
+    """
+    Curved (Superconducting) picture-frame geometry parameterisation, with
+    top and bottoms domed
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1379]: Argument description

File: bluemira/geometry/parameterisations.py [hunk 27]
```diff
@@ -1189,43 +1336,633 @@ def create_shape(self, label=""):
...
+            flip=True,
+            r_c=0,
+        )
+
+        wires.append(bot_leg_curve)
+
+        return BluemiraWire(wires, label=label)
+
+
+class TopDomeFlatInnerCurvedPictureFrame(GeometryParameterisation):
```
Comment (sebkahn) [inline bluemira/geometry/parameterisations.py:1474]: Very long name too

**Review by sebkahn — CHANGES_REQUESTED** [review]

**Comment by sebkahn** [comment]
Hello, nice work!

Main comments: 
 - `PictureFrameTools` is not a class but rather a namespace
 - Description improvement can be made in general
 - Geometry class name are very long, but I suppose it is ok for a pre-refactor

File: bluemira/geometry/parameterisations.py [hunk 28]
```diff
@@ -986,6 +986,191 @@ def _make_control_points(p0, p3, theta0, theta3, l_start, l_end):
         return p1, p2
 
 
+class PictureFrameTools:
```
[the remark this answers is not in the document]
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:989]: It's a mixin class of sorts, which is what I intended. Not sure if you want me to do anything here....

File: bluemira/geometry/parameterisations.py (hunk 5 above)
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:998]: updated

File: bluemira/geometry/parameterisations.py (hunk 6 above)
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:1017]: The description describes the shapes being parameterised, adding detail to the 'returns' section would be repeating this info no?

File: bluemira/geometry/parameterisations.py (hunk 25 above)
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:1376]: Have you got any suggestions? I have to keep it easily understood and not TF coil specific

File: bluemira/geometry/parameterisations.py (hunk 26 above)
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:1379]: So I just followed the same format as all the other parameterisations

File: bluemira/geometry/parameterisations.py (hunk 27 above)
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:1474]: Same point as earlier

File: bluemira/geometry/parameterisations.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (shaildesai96) [reply bluemira/geometry/parameterisations.py:989]: Also namespaces don't exist in Python right? (I think)

File: bluemira/geometry/parameterisations.py (hunk 25 above)
Comment (je-cook) [reply bluemira/geometry/parameterisations.py:1376]: Talked about in the meeting just now. Fine for now will hopefully refactor at some point (allowing PictureFrame to be one class)

File: bluemira/geometry/parameterisations.py (hunk 26 above)
Comment (sebkahn) [reply bluemira/geometry/parameterisations.py:1416]: Bluemira is using the Sphinx doc convention and here is an example of the a class doc in sphinx

image

File: bluemira/geometry/parameterisations.py (hunk 25 above)
Comment (sebkahn) [reply bluemira/geometry/parameterisations.py:1413]: Yes, this is ok.

I just wanted to make these names and class inheritance structure does not stay in the long term.

**Comment by sebkahn** [comment]
Thanks for the update @shaildesai96, still some very easy to comments to be address and we are good to go.

File: bluemira/geometry/parameterisations.py (hunk 26 above)
Comment (CoronelBuendia) [reply bluemira/geometry/parameterisations.py:1416]: The docstring convention we use is described in CONTRIBUTING.md: https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/develop/CONTRIBUTING.md.

File: bluemira/geometry/parameterisations.py (hunk 6 above)
Comment (sebkahn) [reply bluemira/geometry/parameterisations.py:1040]: I suppose I am happy with that.

File: bluemira/geometry/parameterisations.py (hunk 22 above)
Comment (sebkahn) [reply bluemira/geometry/parameterisations.py:996]: These seems to be `np.float` this syntax is then not valid

Comment (je-cook) [reply bluemira/geometry/parameterisations.py:996]: using np.float directly is depreciated you should now just use float. (np.array if its an array)

**Pull request merged**